core/genesis.go

```diff
@@ -208,13 +207,39 @@ func gaFlush(ga *genesisT.GenesisAlloc, db ethdb.Database) (common.Hash, error)
 	}
 	root, err := statedb.Commit(false)
 	if err != nil {
-		return common.Hash{}, err
+		return err
 	}
 	err = statedb.Database().TrieDB().Commit(root, true, nil)
+	if err != nil {
+		return err
+	}
+	// Marshal the genesis state specification and persist.
+	blob, err := json.Marshal(ga)
+	if err != nil {
+		return err
+	}
+	rawdb.WriteGenesisStateSpec(db, root, blob)
+	return nil
+}
+
+// gaDeriveHash computes the state root according to the genesis specification.
+func gaDeriveHash(ga *genesisT.GenesisAlloc) (common.Hash, error) {
```

Comment:
There was (another) refactoring upstream around `Genesis` methods, mostly `Genesis.ToBlock()` and `CommitGenesis` or something. `gaDeriveHash` is our implementation of upstream's `genesisAlloc.deriveHash`; for import cycle reasons its a function here and not a method on the `genesisT#GenesisAlloc` type.